oauthproxy.go

```diff
@@ -1012,7 +891,184 @@ func (p *OAuthProxy) Proxy(rw http.ResponseWriter, req *http.Request) {
 		p.ErrorPage(rw, http.StatusInternalServerError,
 			"Internal Error", "Internal Error")
 	}
+}
+
+// See https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching?hl=en
+var noCacheHeaders = map[string]string{
+	"Expires":         time.Unix(0, 0).Format(time.RFC1123),
+	"Cache-Control":   "no-cache, no-store, must-revalidate, max-age=0",
+	"X-Accel-Expires": "0", // https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/
+}
+
+// prepareNoCache prepares headers for preventing browser caching.
+func prepareNoCache(w http.ResponseWriter) {
+	// Set NoCache headers
+	for k, v := range noCacheHeaders {
+		w.Header().Set(k, v)
+	}
+}
+
+// getOAuthRedirectURI returns the redirectURL that the upstream OAuth Provider will
+// redirect clients to once authenticated.
+// This is usually the OAuthProxy callback URL.
+func (p *OAuthProxy) getOAuthRedirectURI(req *http.Request) string {
+	// if `p.redirectURL` already has a host, return it
+	if p.redirectURL.Host != "" {
+		return p.redirectURL.String()
+	}
+
+	// Otherwise figure out the scheme + host from the request
+	rd := *p.redirectURL
+	rd.Host = requestutil.GetRequestHost(req)
+	rd.Scheme = requestutil.GetRequestProto(req)
+
+	// If CookieSecure is true, return `https` no matter what
+	// Not all reverse proxies set X-Forwarded-Proto
+	if p.CookieSecure {
+		rd.Scheme = schemeHTTPS
+	}
+	return rd.String()
+}
+
+// getAppRedirect determines the full URL or URI path to redirect clients to
+// once authenticated with the OAuthProxy
+// Strategy priority (first legal result is used):
+// - `rd` querysting parameter
+// - `X-Auth-Request-Redirect` header
+// - `X-Forwarded-(Proto|Host|Uri)` headers (when ReverseProxy mode is enabled)
+// - `X-Forwarded-(Proto|Host)` if `Uri` has the ProxyPath (i.e. /oauth2/*)
+// - `X-Forwarded-Uri` direct URI path (when ReverseProxy mode is enabled)
+// - `req.URL.RequestURI` if not under the ProxyPath (i.e. /oauth2/*)
+// - `/`
+func (p *OAuthProxy) getAppRedirect(req *http.Request) (string, error) {
+	err := req.ParseForm()
+	if err != nil {
+		return "", err
+	}
+
+	// These redirect getter functions are strategies ordered by priority
+	// for figuring out the redirect URL.
+	type redirectGetter func(req *http.Request) string
+	for _, rdGetter := range []redirectGetter{
+		p.getRdQuerystringRedirect,
+		p.getXAuthRequestRedirect,
+		p.getXForwardedHeadersRedirect,
+		p.getURIRedirect,
+	} {
+		if redirect := rdGetter(req); redirect != "" {
+			return redirect, nil
+		}
+	}
```

Comment:
This is interesting... It does make the `IsValidRedirect` part of this flow a bit harder to understand though right? 

I think this changes the behaviour as well right, currently we look for the first non empty string, then check if it's valid, then fallback if it's not, now you could have an invalid RD but a valid XAuthRequestRedirect, was that intentional?

Reply:
Let me double check on this one to make sure I got the logic right. I believe the only additional improved logic I added is a final trust of `X-Forwarded-Uri` if set (& trusted) instead of falling back directly to the `req.URL.RequestURI` as the last case scenario in `getURIRedirect`.

All the helper methods should call `IsValidRedirect` internally in their decision branching and return `""` if nothing was valid (and we should fall through).

Do you think they should return a `valid` bool? Or maybe we explicitly do the final `IsValidRedirect` check here instead of `""` to make that contract clear?

Reply:
This was my best attempt to clean up how big and branchy this function got by splitting the various redirect strategies into their own methods. Gotta appease the CodeClimate gods 😄

Reply:
>I think this changes the behaviour as well right, currently we look for the first non empty string, then check if it's valid, then fallback if it's not, now you could have an invalid RD but a valid XAuthRequestRedirect, was that intentional?

Ah - I understand this point better. Yes, I think I made that intentional now that this processing has grown so complex in logic. If user specified variants that are top priority (e.g. `rd`) are illegal (bad formatting, or open redirect attack) we don't just fail to `/`. Now we still fall back to other valid means based in the strategy priority to generate the redirect (nginx header, traefik headers, req URI) if any of those would result in a valid redirect.

Reply:
I wonder if we should log any invalid redirect URIs that we reject higher up in the chain for security logging before falling back to the next priority strategy 🤔

Reply:
Ok I'm cool with this change, think it's the right way to go.

My only two points would be to put an important note into the changelog and as you've suggested, log any invalid redirects further up the chain (but only if they are not the empty string)